src/transformers/tokenization_utils.py

```diff
@@ -408,51 +418,62 @@ def _add_tokens(self, new_tokens: Union[List[str], List[AddedToken]], special_to
         # Note: resize_token_embeddings expects to receive the full size of the new vocabulary, i.e. the length of the tokenizer.
         model.resize_token_embeddings(len(tokenizer))
         ```"""
-        new_tokens = [str(tok) for tok in new_tokens]
-
-        tokens_to_add = []
+        added_tokens = 0
         for token in new_tokens:
-            if not isinstance(token, str):
+            if not isinstance(token, (str, AddedToken)):
                 raise TypeError(f"Token {token} is not a string but a {type(token)}.")
-            if not special_tokens and hasattr(self, "do_lower_case") and self.do_lower_case:
-                token = token.lower()
-            if (
-                token != self.unk_token
-                and self.convert_tokens_to_ids(token) == self.convert_tokens_to_ids(self.unk_token)
-                and token not in tokens_to_add
-            ):
-                tokens_to_add.append(token)
-                if self.verbose:
-                    logger.info(f"Adding {token} to the vocabulary")
-
-        added_tok_encoder = {tok: len(self) + i for i, tok in enumerate(tokens_to_add)}
-        added_tok_decoder = {v: k for k, v in added_tok_encoder.items()}
-        self.added_tokens_encoder.update(added_tok_encoder)
-        self.added_tokens_decoder.update(added_tok_decoder)
-
-        # Make sure we don't split on any special tokens (even they were already in the vocab before e.g. for Albert)
-        if special_tokens:
-            if len(new_tokens) == 1:
-                _insert_one_token_to_ordered_list(self.unique_no_split_tokens, new_tokens[0])
-            else:
-                self.unique_no_split_tokens = sorted(set(self.unique_no_split_tokens).union(set(new_tokens)))
-        else:
-            # Or on the newly added tokens
-            if len(tokens_to_add) == 1:
-                _insert_one_token_to_ordered_list(self.unique_no_split_tokens, tokens_to_add[0])
+            if isinstance(token, str):
+                token = AddedToken(token)
+            if token.content == self.unk_token:
+                self.unk_token = token
+                # if we are initializing token should still be added!
+                continue
+            # if unk_token is not part of the vocab, but we are adding tokens
+            if self.unk_token_id is not None and self.convert_tokens_to_ids(token.content) == self.unk_token_id:
+                # if some tokens were added at the beginning ignore them HACK
+                new_idx = len(self)
+                if not special_tokens and token.normalized and hasattr(self, "do_lower_case") and self.do_lower_case:
+                    # Since we are adding a token to the vocab, let's be consistent: thje vocab probably does not contain any upper-case words
```

Comment:
```suggestion
                    # Since we are adding a token to the vocab, let's be consistent: the vocab probably does not contain any upper-case words
```